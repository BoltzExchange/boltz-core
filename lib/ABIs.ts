import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
