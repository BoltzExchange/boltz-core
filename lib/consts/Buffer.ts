export const Nonce = Buffer.from('00', 'hex');
export const EmptyScript = Buffer.alloc(0);
export const PrefixUnconfidential = Buffer.from('01', 'hex');
