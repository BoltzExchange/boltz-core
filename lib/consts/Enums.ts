export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
