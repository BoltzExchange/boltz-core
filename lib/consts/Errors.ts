const Errors = {
  TAPROOT_NOT_SUPPORTED: 'taproot not supported',
};

export default Errors;
