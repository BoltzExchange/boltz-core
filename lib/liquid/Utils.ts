import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
