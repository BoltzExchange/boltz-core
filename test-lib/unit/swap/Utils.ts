import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
