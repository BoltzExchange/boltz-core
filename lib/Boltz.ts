import Networks from './consts/Networks';
import * as Scripts from './swap/Scripts';
import { OutputType } from './consts/Enums';
import * as SwapUtils from './swap/SwapUtils';
import { swapScript } from './swap/SwapScript';
import { detectSwap } from './swap/SwapDetector';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import { estimateFee, estimateSize } from './FeeCalculator';
import { ScriptElement, TransactionOutput } from './consts/Types';

export {
  Networks,
  OutputType,
  ScriptElement,
  TransactionOutput,

  Scripts,

  swapScript,

  detectSwap,
  detectPreimage,

  estimateFee,
  estimateSize,

  constructClaimTransaction,
  constructRefundTransaction,

  SwapUtils,
};
