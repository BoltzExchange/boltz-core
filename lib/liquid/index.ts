import { init } from './Init';
import Errors from './consts/Errors';
import Networks from './consts/Networks';
import { getOutputValue } from './Utils';
import { constructClaimTransaction } from './swap/Claim';
import { constructRefundTransaction } from './swap/Refund';

export {
  Errors,
  Networks,
  init,
  getOutputValue,
  constructClaimTransaction,
  constructRefundTransaction,
};
