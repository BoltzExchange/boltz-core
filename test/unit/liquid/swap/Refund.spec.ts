import { confidential } from 'liquidjs-lib';
import zkp from '@michael1011/secp256k1-zkp';
import { ECPair } from '../../Utils';
import { lbtcRegtest, nonce } from './ClaimDetails';
import { getHexBuffer } from '../../../../lib/Utils';
import { OutputType } from '../../../../lib/consts/Enums';
import { LiquidRefundDetails } from '../../../../lib/liquid';
import { constructRefundTransaction, init } from '../../../../lib/liquid';

describe('Liquid Refund', () => {
  const utxo = {
    nonce,
    vout: 0,
    asset: lbtcRegtest,
    type: OutputType.Bech32,
    keys: ECPair.fromPrivateKey(
      getHexBuffer(
        '96960cf685a4c1f174ff0918ba937cd2d5090c0d7e20646b091b173e55dadb52',
      ),
    ),
    value: confidential.satoshiToConfidentialValue(2000),
    script: getHexBuffer(
      '00206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9e',
    ),
    txHash: getHexBuffer(
      '285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d754',
    ),
    redeemScript: getHexBuffer(
      'a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac',
    ),
  };

  const testRefund = (utxos: LiquidRefundDetails[], fee: number) => {
    return constructRefundTransaction(
      utxos,
      getHexBuffer('00140000000000000000000000000000000000000000'),
      11,
      fee,
    );
  };

  beforeAll(async () => {
    init(await zkp());
  });

  test('should refund a P2WSH swap', async () => {
    expect(testRefund([utxo], 21).toHex()).toMatchSnapshot();
  });

  test('should refund multiple swaps in one transaction', () => {
    expect(
      testRefund(
        [
          utxo,
          {
            ...utxo,
            vout: utxo.vout + 1,
          },
          {
            ...utxo,
            vout: utxo.vout + 2,
          },
        ],
        38,
      ).virtualSize(),
    ).toMatchSnapshot();
  });
});
