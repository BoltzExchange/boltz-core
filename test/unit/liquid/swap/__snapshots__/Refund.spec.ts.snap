// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liquid Refund should refund a P2WSH swap 1`] = `"020000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000fdffffff02016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f0100000000000007bb001600140000000000000000000000000000000000000000016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f01000000000000001500000b00000000000347304402201fae80174ca581816625c29a417e5fc85ba6fab31037f9e167e0d7619115e419022012833d8fd763dd1592a4cfd931a859faaa87feb146f8db12f513b114d8c96fee010064a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0000000000"`;

exports[`Liquid Refund should refund multiple swaps in one transaction 1`] = `379`;
